### PR TITLE
[sgd] fp16 (apex) and scheduler support + move examples page

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Transparently support apex for mixed precision training. Also support learning rate schedulers.

TODO:

1. Documentation
  - [x] Mixed Precision
  - [x] Custom Schedulers
2. Extensive Tests.
  - [x] run tests
  - [x] run distributed tests

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# flake8: noqa
+"""
+This file holds code for the Torch best-practices guide in the documentation.
```
Comment (richardliaw) [inline doc/examples/doc_code/raysgd_torch_signatures.py]: edit

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    """Constructor of one or more Torch optimizer schedulers.
+
+    Note that optionally, you can pass in a Torch Scheduler constructor directly
+    into the PyTorchTrainer. For example:
+
+    .. code-block:: python
+
+        PyTorchTrainer(
+            scheduler_creator=torch.optim.lr_scheduler.ReduceLROnPlateau,
+            ...)``.
```
Comment (richardliaw) [inline doc/examples/doc_code/raysgd_torch_signatures.py]: remove

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 3]
```diff
@@ -223,7 +207,47 @@ The trained torch model can be extracted for use within the same Python program
...
+    trainer = PyTorchTrainer(
+        model_creator,
+        data_creator,
+        optimizer_creator,
+        loss_creator=nn.MSELoss,
+        num_replicas=4,
+        use_fp16=True
+    )
+
+``Apex`` is a Pytorch extension with NVIDIA-maintained utilities to streamline mixed precision and distributed training. When using ``use_fp16``, you should not manually cast their model or data to .half(). The flag informs the Trainer to call ``amp.initialize`` on the provided models and optimizers underneath the hood, along with using the scaled loss: ``amp.scale_loss(loss, optimizer)``.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_pytorch.rst]: edit

**Comment by richardliaw** [comment]
Right now, this PR breaks training - 

```
# regular 8 gpus, 20 epochs
'mean_accuracy': 0.2263, 'mean_loss': 2.153756123352051

# fp16 8 gpus, 20 epochs
'mean_accuracy': 0.8548, 'mean_loss': 0.45528498129844663 
```

File: python/ray/experimental/sgd/pytorch/examples/train_example.py [hunk 4]
```diff
@@ -40,6 +40,11 @@ def optimizer_creator(model, config):
     return torch.optim.SGD(model.parameters(), lr=1e-2)
 
 
+def scheduler_creator(optimizer, config):
+    """Returns optimizer."""
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/examples/train_example.py]: fix

**Comment by richardliaw** [comment]
jenkins test tune

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+# flake8: noqa
+"""
+This file holds code for the Pytorch Trainer creator signatures
```
Comment (richardliaw) [inline doc/examples/doc_code/raysgd_torch_signatures.py]:
```suggestion
This file holds code for the Pytorch Trainer creator signatures.
```

File: doc/source/index.rst [hunk 6]
```diff
@@ -280,7 +280,6 @@ Getting Involved
 
    raysgd/raysgd.rst
    raysgd/raysgd_pytorch.rst
-   raysgd/raysgd_pytorch_examples.rst
```
Comment (richardliaw) [inline doc/source/index.rst]: removed explicit examples page.

File: doc/source/raysgd/raysgd.rst [hunk 7]
```diff
@@ -1,10 +1,6 @@
 RaySGD: Distributed Deep Learning
 =================================
 
-.. tip:: Get in touch with us if you're using or considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
-.. warning:: This is still an experimental API and is subject to change in the near future.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd.rst]: Warning is now moved to subsections of the doc.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 8]
```diff
@@ -1,135 +1,125 @@
 RaySGD Pytorch
 ==============
 
-.. warning:: This is still an experimental API and is subject to change in the near future.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_pytorch.rst]: (moved to subsections where API is actually more experimental)

**Comment by richardliaw** [comment]
cc @wuisawesome

File: python/ray/experimental/sgd/pytorch/utils.py [hunk 9]
```diff
@@ -45,42 +97,85 @@ def train(model, train_iterator, criterion, optimizer, config):
...
+
+    This function automatically measures timing for various operations such
+    as host to device transfer and processing time for the batch.
+
+    It also automatically detects and places the data on the given GPU device
+    if available.
+
+    Raises:
+        ValueError if multiple models/schedulers are provided. You
+            are expected to have a custom training function if you wish
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/utils.py]:
```suggestion
            are expected to have a custom validation function if you wish
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 10]
```diff
@@ -223,7 +206,48 @@ The trained torch model can be extracted for use within the same Python program
...
+        optimizer_creator,
+        loss_creator=nn.MSELoss,
+        num_replicas=4,
+        use_fp16=True
+    )
+
+``Apex`` is a Pytorch extension with NVIDIA-maintained utilities to streamline mixed precision and distributed training. When ``use_fp16=True``,
+you should not manually cast your model or data to ``.half()``. The flag informs the Trainer to call ``amp.initialize`` on the created models and optimizers and optimize using the scaled loss: ``amp.scale_loss(loss, optimizer)``.
+
+To specify particular parameters for ``amp.initialize``, you can use the ``apex_args`` field for the PyTorchTrainer constructor:
```
Comment (wuisawesome) [inline doc/source/raysgd/raysgd_pytorch.rst]: Should there be a link/explanation of what to pass in for apex_args? Maybe we can add a link to this: https://nvidia.github.io/apex/amp.html#module-apex.amp

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 11]
```diff
@@ -396,11 +436,56 @@ Note that this is needed if the model creator returns multiple models.
...
+
+Have features that you'd really like to see in RaySGD? Feel free to `open an issue <https://github.com/ray-project/ray>`_.
+
+.. _raysgd-pytorch-examples:
+
+PyTorchTrainer Examples
+-----------------------
+
+Here are some examples of using RaySGD for training PyTorch models. If you'd like
+to contribute an example, feel free to create a `pull request here <https://github.com/ray-project/ray/pull/>`_.
```
Comment (wuisawesome) [inline doc/source/raysgd/raysgd_pytorch.rst]: Link broken for me, try this? https://github.com/ray-project/ray/pull/new/master

File: python/ray/experimental/sgd/pytorch/examples/cifar_pytorch_example.py [hunk 12]
```diff
@@ -96,23 +48,33 @@ def optimizer_creator(model, config):
...
         use_gpu=use_gpu,
         batch_size=16 if test_mode else 512,
-        backend="nccl" if use_gpu else "gloo")
-    for i in range(5):
-        stats = trainer1.train()
+        backend="nccl" if use_gpu else "gloo",
+        scheduler_step_freq="epoch",
+        use_fp16=use_fp16)
+    for i in range(num_epochs):
+        stats = trainer1.train(max_retries=0)
```
Comment (wuisawesome) [inline python/ray/experimental/sgd/pytorch/examples/cifar_pytorch_example.py]: Is there a reason for explicitly specifying the default option here?

File: python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py [hunk 13]
```diff
@@ -109,26 +116,15 @@ def get_state(self):
...
+    def _set_model_state_dicts(self, model_state_dicts):
+        for model, model_state_dict in zip(self.models, model_state_dicts):
             model.module.load_state_dict(model_state_dict)
-        for optimizer, opt_state_dict in zip(self.optimizers,
-                                             state["optimizers"]):
-            optimizer.load_state_dict(opt_state_dict)
-        self.epoch = state["stats"]["epoch"]
 
-    def shutdown(self):
+    # def shutdown(self):
```
Comment (wuisawesome) [inline python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]: Unsure if this is worth deprecating/warning instead of removing?

File: python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py [hunk 14]
```diff
@@ -98,8 +105,8 @@ def step(self):
             self.train_loader.sampler.set_epoch(self.epoch)
         return super(DistributedPyTorchRunner, self).step()
 
-    def get_state(self):
-        """Returns the state of the runner."""
+    def _get_model_state_dicts(self):
```
Comment (wuisawesome) [inline python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]: Aren't `_get_mode_state_dicts` and `_set_model_state_dicts` identical to the parent class implementation now?

File: python/ray/experimental/sgd/pytorch/examples/example-sgd.yaml [hunk 15]
```diff
@@ -37,18 +36,18 @@ head_node:
 worker_nodes:
     InstanceType: p3.8xlarge
     ImageId: ami-0757fc5a639fe7666
+    # Run workers on spot by default. Comment this out to use on-demand.
     InstanceMarketOptions:
         MarketType: spot
```
Comment (wuisawesome) [inline python/ray/experimental/sgd/pytorch/examples/example-sgd.yaml]: perhaps a comment somewhere that using spot instances and `max_retries=0` could end poorly

File: python/ray/experimental/sgd/pytorch/examples/train_example.py [hunk 16]
```diff
@@ -40,6 +40,11 @@ def optimizer_creator(model, config):
     return torch.optim.SGD(model.parameters(), lr=1e-2)
 
 
+def scheduler_creator(optimizer, config):
```
Comment (wuisawesome) [inline python/ray/experimental/sgd/pytorch/examples/train_example.py]: i might be off about something here, but should this example also show how to call `update_scheduler`?

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    You will also need to provide a custom training
+    function to specify the optimization procedure for multiple models.
+
+    Args:
+        config (dict): Configuration dictionary passed into ``PyTorchTrainer``.
+
+    Returns:
+        One or more torch.nn.Module objects.
+    """
+    return nn.Linear(1, 1)
```
Comment (edoakes) [inline doc/examples/doc_code/raysgd_torch_signatures.py]: Is this supposed to completely ignore `config`?

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    return torch.optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.9)
+
+# __torch_scheduler_end__
+
+# __torch_ray_start__
+import ray
+
+ray.init()
+# or ray.init(address="auto") if a cluster has been started.
```
Comment (edoakes) [inline doc/examples/doc_code/raysgd_torch_signatures.py]:
```suggestion
# or ray.init(address="auto") to connect to a running cluster.
```

File: doc/examples/doc_code/raysgd_torch_signatures.py (hunk 17 above)
Comment (edoakes) [reply doc/examples/doc_code/raysgd_torch_signatures.py]: I see, `config` is also user-specified and this is just a simple example that hardcodes the model creator. Might help to either set a trivial config option or else leave a comment as to why it's hardcoded (because it's a toy example).

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+
+    Args:
+        optimizers: The return values from ``optimizer_creator``.
+            This can be one or more torch optimizer objects.
+        config: Configuration dictionary passed into ``PyTorchTrainer``
+
+    Returns:
+        One or more Torch scheduler objects.
+    """
+    return torch.optim.lr_scheduler.StepLR(optimizer, step_size=5, gamma=0.9)
```
Comment (edoakes) [inline doc/examples/doc_code/raysgd_torch_signatures.py]:
```suggestion
    return torch.optim.lr_scheduler.StepLR(optimizers, step_size=5, gamma=0.9)
```
Might want to double check that the code samples run

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 20]
```diff
@@ -1,135 +1,124 @@
...
 
-.. warning:: This is still an experimental API and is subject to change in the near future.
-
-.. tip:: Get in touch with us if you're using or considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
-
 .. image:: raysgd-pytorch.svg
     :align: center
 
-The RaySGD ``PyTorchTrainer`` simplifies distributed model training for PyTorch. The ``PyTorchTrainer`` is a wrapper around ``torch.distributed.launch`` with a Python API to easily incorporate distributed training into a larger Python application, as opposed to needing to execute training outside of Python. For end to end examples, see :ref:`raysgd-pytorch-example`.
+The RaySGD ``PyTorchTrainer`` simplifies distributed model training for PyTorch. The ``PyTorchTrainer`` is a wrapper around ``torch.distributed.launch`` with a Python API to easily incorporate distributed training into a larger Python application, as opposed to needing to execute training outside of Python.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: What exactly does "execute training outside of python" mean? This motivation could be stronger if it was a bit more concrete (e.g., "without needing to wrap your training code in bash scripts).

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 21]
```diff
@@ -1,135 +1,124 @@
...
+For end to end examples leveraging RaySGD PyTorchTrainer, jump to :ref:`raysgd-pytorch-examples`.
 
 Setting up training
 -------------------
 
-The ``PyTorchTrainer`` can be constructed with functions that wrap components of the training script. Specifically, it needs constructors for the Model, Data, Optimizer, and Loss to create replicated copies across different devices and machines.
+.. tip:: We need your feedback! RaySGD is currently early in its development, and we're hoping to get feedback from people using or considering it. We'd love `to get in touch <https://forms.gle/26EMwdahdgm7Lscy9>`_!
 
-For example:
+The ``PyTorchTrainer`` can be constructed with functions that wrap components of the training script. Specifically, it needs constructors for the Model, Data, Optimizer, Loss, and ``lr_scheduler`` to create replicated copies across different devices and machines.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
The ``PyTorchTrainer`` can be constructed with functions that wrap components of the training script. Specifically, it requires constructors for the Model, Data, Optimizer, Loss, and ``lr_scheduler`` to create replicated copies across different devices and machines.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 22]
```diff
@@ -1,135 +1,124 @@
...
-    ray.init()
-    # or ray.init(address="auto") if a cluster has been started.
+.. _starting-pytorch-trainer:
 
-Instantiate the trainer object:
+Putting things together
+~~~~~~~~~~~~~~~~~~~~~~~
 
-.. code-block:: python
+Before instantiating the trainer, you'll have to start or connect to a Ray cluster:
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Before instantiating the trainer, first start or connect to a Ray cluster:
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 23]
```diff
@@ -154,17 +143,19 @@ You can customize the exact function that is called by using a customized traini
...
 
-After training, you may want to reappropriate the Ray cluster. To release Ray resources obtained by the trainer:
+After training, you may want to reappropriate the Ray cluster. To release Ray resources obtained by the Trainer:
 
 .. code-block:: python
 
     trainer.shutdown()
 
-.. note:: Be sure to call ``save`` or ``get_model`` before shutting down.
+.. note:: Be sure to call ``trainer.save()`` or ``trainer.get_model()`` before shutting down.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: Maybe just put this in the code block to make sure people don't miss it?

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 24]
```diff
@@ -154,17 +143,19 @@ You can customize the exact function that is called by using a customized traini
...
 
 Initialization Functions
 ------------------------
 
+.. warning:: This is still an experimental API and is subject to change in the near future.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
.. warning:: This is still an experimental API and is subject to change without warning.
```

Comment (edoakes) [reply doc/source/raysgd/raysgd_pytorch.rst]: unless it isn't

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 25]
```diff
@@ -223,7 +206,48 @@ The trained torch model can be extracted for use within the same Python program
 .. code-block::
 
     trainer.train()
-    model = trainer.get_model()
+    model = trainer.get_model()  # will return multiple models if the model_creator does so.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
    model = trainer.get_model()  # will return multiple models if the model_creator does so.
```
```suggestion
    model = trainer.get_model()  # Returns multiple models if the model_creator does.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 26]
```diff
@@ -223,7 +206,48 @@ The trained torch model can be extracted for use within the same Python program
...
+
+Mixed Precision (FP16) Training
+-------------------------------
+
+You can enable mixed precision training for PyTorch by the ``use_fp16`` flag. This automatically converts the model(s) and optimizer(s) to train using mixed-precision. This requires NVIDIA ``Apex``, which can be installed from `the NVIDIA/Apex repository <https://github.com/NVIDIA/apex#quick-start>`_:
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
You can enable mixed precision training for PyTorch with the ``use_fp16`` flag. This automatically converts the model(s) and optimizer(s) to train using mixed-precision. This requires NVIDIA ``Apex``, which can be installed from `the NVIDIA/Apex repository <https://github.com/NVIDIA/apex#quick-start>`_:
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 27]
```diff
@@ -236,17 +260,17 @@ You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ o
 .. code-block:: bash
 
     ray up CLUSTER.yaml
-    python train.py --address="auto"
+    ray submit train.py --args="--address='auto'"
 
-Then, you'll be able to scale up the number of workers seamlessly across multiple nodes:
+Then, within ``train.py`` you'll be able to scale up the number of workers seamlessly across multiple nodes:
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Then, within ``train.py`` you can scale up the number of workers seamlessly across multiple nodes:
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 28]
```diff
@@ -281,19 +305,21 @@ Users can set ``checkpoint="auto"`` to always checkpoint the current model befor
 Advanced: Hyperparameter Tuning
 -------------------------------
 
+.. warning:: This is still an experimental API and is subject to change in the near future.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: ditto

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 29]
```diff
@@ -336,24 +362,38 @@ You can see the `DCGAN script <https://github.com/ray-project/ray/blob/master/py
...
 
-Note that this is needed if the model creator returns multiple models.
+Note that this is needed if the model creator returns multiple models, optimizers, or schedulers.
 
 .. code-block:: python
 
-    def train(models, dataloader, criterion, optimizers, config):
-        """A custom training function.
+    # An example custom training function.
+
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: remove newline?

Comment (edoakes) [reply doc/source/raysgd/raysgd_pytorch.rst]: also consider removing the comment, doesn't add much on top of the name of the section

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 30]
```diff
@@ -396,11 +436,56 @@ Note that this is needed if the model creator returns multiple models.
...
+                wraps the provided Dataset.
+            criterion: The loss object created by the loss_creator.
+            scheduler (optional): The torch.optim.lr_scheduler object(s)
+                as created by the scheduler_creator.
+
+        Returns:
+            A dict of metrics from the evaluation.
+        """
+        ...
+        return {...}
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: Consider adding some dummy common values here

**Review by edoakes — COMMENTED** [review]
Only looked at docs so far. Review of code inbound soon

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py [hunk 31]
```diff
@@ -91,6 +111,21 @@ def _create_loss(self):
         if torch.cuda.is_available():
             self.criterion = self.criterion.cuda()
 
+    def _create_schedulers_if_available(self):
+        if not self.scheduler_creator:
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/pytorch_runner.py]: Why is this necessary? Can the user add the scheduler after initializing the `DistributedPyTorchRunner`? Probably worth a comment.

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py [hunk 32]
```diff
@@ -177,20 +224,44 @@ def get_state(self):
             for k, v in state_dict.items():
                 state_dict[k] = v.cpu()
             cpu_state_dicts += [state_dict]
-        return {
+        return cpu_state_dicts
+
+    def _set_model_state_dicts(self, models_state_dicts):
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/pytorch_runner.py]: Should this be `_load_model_state_dicts` to match the call below?

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py [hunk 33]
```diff
@@ -177,20 +224,44 @@ def get_state(self):
             for k, v in state_dict.items():
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/pytorch_runner.py]: This whole loop body could be a dict comprehension:
```
cpu_state_dicts.append({k: v.cpu() for k, v in model.state_dict().items()})
```

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py [hunk 34]
```diff
@@ -177,20 +224,44 @@ def get_state(self):
...
             "optimizers": [opt.state_dict() for opt in self.optimizers],
             "stats": self.stats()
         }
+        if self.schedulers:
+            state.update({
+                "schedulers": [
+                    scheduler.state_dict() for scheduler in self.schedulers
+                ]
+            })
+        if self.use_fp16 and amp:
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/pytorch_runner.py]: Maybe comment what this check is for or factor it out into a util function? Seems like I've seen it at least twice.

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py [hunk 35]
```diff
@@ -147,9 +169,25 @@ def __init__(self,
         self.use_gpu = use_gpu
         self.batch_size = batch_size
         self.max_replicas = num_replicas
+
+        self.use_fp16 = use_fp16
+
+        if apex_args and type(apex_args) is not dict:
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: maybe use `isinstance` instead so people could pass a defaultdict or other types

File: python/ray/experimental/sgd/pytorch/utils.py [hunk 36]
```diff
@@ -4,32 +4,84 @@
 
 from ray.experimental.sgd.utils import TimerStat
 
-
-def train(model, train_iterator, criterion, optimizer, config):
-    """Runs 1 training epoch"""
+amp = None
+
+try:
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/utils.py]: Maybe leave a comment on why this is optional/what doesn't work if it isn't present?

File: python/ray/experimental/sgd/pytorch/utils.py [hunk 37]
```diff
@@ -45,42 +97,85 @@ def train(model, train_iterator, criterion, optimizer, config):
...
+                loss.backward()
 
         with timers["apply"]:
             # Call step of optimizer to update model params
             optimizer.step()
 
+        if scheduler and config.get(SCHEDULER_STEP) == SCHEDULER_STEP_BATCH:
+            scheduler.step()
+
         # measure elapsed time
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/utils.py]: not you but punctuation

File: python/ray/experimental/sgd/pytorch/utils.py [hunk 38]
```diff
@@ -45,42 +97,85 @@ def train(model, train_iterator, criterion, optimizer, config):
...
+        model: The model as created by the model_creator.
+        train_iterator: An iterator created from the DataLoader which
+            wraps the provided Dataset.
+        criterion: The loss object created by the loss_creator.
+        scheduler (optional): The torch.optim.lr_scheduler object
+            as created by the scheduler_creator. By default,
+            this is not used in this function.
+
+    Returns:
+        A dict of metrics from the evaluation.
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/utils.py]: Does the format of this dictionary matter?

File: python/ray/experimental/sgd/pytorch/utils.py [hunk 39]
```diff
@@ -99,8 +194,15 @@ def validate(model, val_iterator, criterion, config):
...
-    stats.update(mean_accuracy=correct / total)
+            if config.get(TEST_MODE) and batch_idx == 0:
+                break
+
+    stats = {
+        BATCH_COUNT: batch_idx + 1,
+        "batch_time": batch_time.avg,
+        "validation_loss": losses.avg
+    }
+    stats.update(mean_accuracy=correct / total, mean_loss=losses.sum / total)
```
Comment (edoakes) [inline python/ray/experimental/sgd/pytorch/utils.py]: why not just initialize these inline above?

**Review by edoakes — COMMENTED** [review]
LGTM overall just some small suggestions

File: doc/examples/doc_code/raysgd_torch_signatures.py [hunk 40]
```diff
@@ -0,0 +1,121 @@
+# flake8: noqa
```
Comment (richardliaw) [inline doc/examples/doc_code/raysgd_torch_signatures.py]: Add this to jenkins

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 20 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
The RaySGD ``PyTorchTrainer`` simplifies distributed model training for PyTorch. The ``PyTorchTrainer`` is a wrapper around ``torch.distributed.launch`` with a Python API to easily incorporate distributed training into a larger Python application, as opposed to needing to wrap your training code in bash scripts.
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 23 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: Think the way it is right now (in a highlighted block of text), it'll be OK because there is a specific section on saving/getting the model.

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 28 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
.. warning:: This is still an experimental API and is subject to change without warning.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 41]
```diff
@@ -336,24 +362,38 @@ You can see the `DCGAN script <https://github.com/ray-project/ray/blob/master/py
...
 ``PyTorchTrainer`` allows you to run a custom training and validation step in parallel on each worker, providing a flexibility similar to using PyTorch natively. This is done via the ``train_function`` and ``validation_function`` parameters.
 
-Note that this is needed if the model creator returns multiple models.
+Note that this is needed if the model creator returns multiple models, optimizers, or schedulers.
 
 .. code-block:: python
 
-    def train(models, dataloader, criterion, optimizers, config):
-        """A custom training function.
+    # An example custom training function.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 30 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
        return {"validation_accuracy": 0.5}
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 11 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: Ah, unfortunately it probably makes more sense to point them to the contributions page - just pointing them to repo for now

```suggestion
to contribute an example, feel free to create a `pull request here <https://github.com/ray-project/ray/>`_.
```

File: python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py [hunk 42]
```diff
@@ -98,8 +105,8 @@ def step(self):
...
+        """Override the parent class method for supporting model.module."""
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]:
```suggestion
        """Fetch state from ``model.module``, needed for PyTorch DDP."""
```

File: python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py (hunk 14 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]: Added a better docstring

Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]: "Fetch state from ``model.module``, which is needed for PyTorch DDP."

File: python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py (hunk 13 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/distributed_pytorch_runner.py]: Ah, on master, it just calls the super class already, so removing it doesn't change behavior.

File: python/ray/experimental/sgd/pytorch/examples/cifar_pytorch_example.py (hunk 12 above)
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/examples/cifar_pytorch_example.py]:
```suggestion
        # Increase `max_retries` to turn on fault tolerance.
        stats = trainer1.train(max_retries=0)
```

Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/examples/cifar_pytorch_example.py]: Thanks - made it more explicit to point out to users when to change `max_retries`.

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py (hunk 35 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_trainer.py]:
```suggestion
        if apex_args and not isinstance(apex_args, dict):
```

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py (hunk 34 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_runner.py]:
```suggestion
        # Check if fp16 is True and if NVIDIA Apex is imported.
        if self.use_fp16 and amp:
```

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py (hunk 31 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_runner.py]:
```suggestion
        # Learning rate schedules are optional.
        if not self.scheduler_creator:
```

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py (hunk 32 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_runner.py]: I think not - the trainer naming symmetry is "get_model_dict" and "set_model_dict". The underlying framework call is  `state_dict` -> `load_state_dict`.

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 29 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: done

File: python/ray/experimental/sgd/pytorch/utils.py (hunk 39 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/utils.py]: done, good idea

File: python/ray/experimental/sgd/pytorch/utils.py (hunk 38 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/utils.py]: Nope

File: python/ray/experimental/sgd/pytorch/utils.py (hunk 36 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/utils.py]: done, good idea

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py (hunk 33 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_runner.py]: adjusted

File: doc/examples/doc_code/raysgd_torch_signatures.py (hunk 19 above)
Comment (richardliaw) [reply doc/examples/doc_code/raysgd_torch_signatures.py]: done, this was a very good suggestion, thanks edward

File: doc/examples/doc_code/raysgd_torch_signatures.py (hunk 17 above)
Comment (richardliaw) [reply doc/examples/doc_code/raysgd_torch_signatures.py]: done, good idea

File: python/ray/experimental/sgd/pytorch/examples/train_example.py (hunk 16 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/examples/train_example.py]: done, added comments

File: python/ray/experimental/sgd/pytorch/examples/example-sgd.yaml (hunk 15 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/examples/example-sgd.yaml]: Hm, I think this belongs not in the cluster specification but rather in the python code? 

Won't do this for now but will keep in mind.

**Review by edoakes — APPROVED** [review]
LGTM

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune
